fixes #11771 - ensure created HostStatuses are saved correctly

domcleal opened the pull request:
The HostStatus reported_at timestamp is now validated to match the DB
schema and populated if it's missing.  HostStatus subclasses should
validate correctly without needing to call #refresh.

Since creation of a ConfigurationStatus etc. can be triggered by an API
render (e.g. through foreman_hooks), the #relevant? method has been
implemented to ensure it remains hidden when not actually in use.

Host::Managed#get_status will now only create a new status object if it
doesn't already exist.  It would only check those already in the DB for
an existing host, so multiple #get_status calls would create a new
status object each time, causing duplicates.  Extra validation and an
index have been added to ensure uniqueness.

Commit:
fixes #11771 - ensure created HostStatuses are saved correctly

The HostStatus reported_at timestamp is now validated to match the DB
schema and populated if it's missing.  HostStatus subclasses should
validate correctly without needing to call #refresh.

Since creation of a ConfigurationStatus etc. can be triggered by an API
render (e.g. through foreman_hooks), the #relevant? method has been
implemented to ensure it remains hidden when not actually in use.

Host::Managed#get_status will now only create a new status object if it
doesn't already exist.  It would only check those already in the DB for
an existing host, so multiple #get_status calls would create a new
status object each time, causing duplicates.  Extra validation and an
index have been added to ensure uniqueness.

domcleal commented:
@stbenjam would you mind testing if this fixes the problem you hit too?  It seems to fix host creation when foreman_hooks is in use with a host/managed/create/ hook, which was rendering the API view, calling Host::Managed#configuration_status and triggering creation of a status.

stbenjam commented:
@domcleal I do indeed have some hooks

```
.
└── host
    └── managed
        ├── after_commit
        │   └── hook_test.sh
        └── after_update
            └── foo.sh
```

However, now when I edit a host, I get some craziness:

![um](https://cloud.githubusercontent.com/assets/429763/11035398/55fdefca-86c0-11e5-8f93-45bd962b1ed5.png)

domcleal commented:
Can you check if you have multiple host statuses in the DB for that one host, or if it's happening at runtime?

It's possible this is present in the DB from a software error as we're lacking indexes and validation to ensure uniqueness.

stbenjam commented:
They are in the DB:

```
2.1.5 :002 > Host.find('rex01.example.com').host_statuses
  Host::Managed Load (0.5ms)  SELECT "hosts".* FROM "hosts" WHERE "hosts"."type" IN ('Host::Managed') AND "hosts"."name" = 'rex01.example.com' LIMIT 1
  HostStatus::Status Load (0.3ms)  SELECT "host_status".* FROM "host_status" WHERE "host_status"."host_id" = 24
+-----+---------------------------------+--------+---------+-------------------------+
| id  | type                            | status | host_id | reported_at             |
+-----+---------------------------------+--------+---------+-------------------------+
| 100 | HostStatus::ConfigurationStatus | 0      | 24      | 2015-11-09 14:00:04 UTC |
| 101 | HostStatus::ConfigurationStatus | 0      | 24      | 2015-11-09 14:00:04 UTC |
| 102 | HostStatus::ConfigurationStatus | 0      | 24      | 2015-11-09 14:00:04 UTC |
| 103 | HostStatus::BuildStatus         | 0      | 24      | 2015-11-09 14:00:04 UTC |
| 104 | HostStatus::ExecutionStatus     | 0      | 24      | 2015-11-09 14:00:04 UTC |
| 105 | HostStatus::BuildStatus         | 0      | 24      | 2015-11-09 14:00:04 UTC |
| 106 | HostStatus::BuildStatus         | 0      | 24      | 2015-11-09 14:00:04 UTC |
| 107 | HostStatus::ExecutionStatus     | 0      | 24      | 2015-11-09 14:00:04 UTC |
| 108 | HostStatus::ExecutionStatus     | 0      | 24      | 2015-11-09 14:00:04 UTC |
+-----+---------------------------------+--------+---------+-------------------------+
```

It only happens with hooks enabled.

stbenjam commented:
I had cleared all the statuses out of the DB before I tested this, so the host had no statuses, I edited it, and then it ended up with 3 of each.  `after_commit` happens in the transaction right? So is it possible it doesn't see the other statuses it created when the hooks run for some reason?

domcleal commented:
The same method that lazily created host status objects was creating additional ones for existing hosts every time it was called during a save.  I've changed that, added an index, validation and a DB migration to clean up the mess.

stbenjam commented:
Much better :smile:

![better](https://cloud.githubusercontent.com/assets/429763/11039678/e068cf02-86d6-11e5-88c5-09dd81ba8f63.png)

domcleal commented:
Thanks for checking.  The last change is just to improve the commit message, no patch changes.

dLobatog commented on test/unit/host_status/build_status_test.rb:
```diff
@@ -7,6 +7,10 @@ def setup
     @status.host = @host
   end
 
+  test 'is valid' do
```
What is this testing exactly?

domcleal replied:
The bug came about because #get_status would create a new ConfigurationStatus and then it would get saved, which failed by default unless you filled in the reported_at field.  I wanted to ensure that calling .new would result in a valid object - so it exercises the before_validation.

dLobatog commented on test/unit/host_status/configuration_status_test.rb:
```diff
@@ -8,7 +8,10 @@ def setup
-    @status.refresh!
+  end
+
+  test 'is valid' do
```
Ditto

dLobatog commented on app/models/host_status/status.rb:
```diff
@@ -48,6 +54,10 @@ def relevant?
 
     private
 
+    def populate_timestamp
```
How about moving this `if` condition to the `before_validation`, and just calling `:update_timestamp` from there?

``` ruby
before_validation :update_timestamp, :unless => ->(host) { host.reported_at.present? }
```

domcleal replied:
That's nice, thanks.  Updated.

dLobatog commented on app/models/host_status/configuration_status.rb:
```diff
@@ -77,6 +77,10 @@ def to_status(options = {})
       end
     end
 
+    def relevant?
```
Kind of unrelated but if you could add a comment on `HostStatus::Status#relevant?` explaining what that method is used for that'd be helpful. It's hard to understand, even in context.

domcleal replied:
Yep, done.

dLobatog commented:
@domcleal Works well w/ a simple after_create hook as the original reporters used. See comments inline, thanks!

domcleal commented:
Thanks, PR updated.

dLobatog commented:
Merged as 55d3360a44e978808c4ffd056c7a79a247fc900b, thanks @domcleal!
